base/subarray.jl

```diff
@@ -3,30 +3,78 @@
 typealias NonSliceIndex Union{Colon, AbstractVector}
 typealias ViewIndex Union{Int, NonSliceIndex}
 
-# LD is the last dimension up through which this object has efficient
-# linear indexing. If LD==length(I), then the object itself has efficient
-# linear indexing.
-immutable SubArray{T,N,P<:AbstractArray,I<:Tuple{Vararg{ViewIndex}},LD} <: AbstractArray{T,N}
+# L is true if the view itself supports fast linear indexing
+immutable SubArray{T,N,P,I,L} <: AbstractArray{T,N}
     parent::P
     indexes::I
     dims::NTuple{N,Int}
-    first_index::Int   # for linear indexing and pointer
+    first_index::Int   # for linear indexing and pointer, only valid when L==true
     stride1::Int       # used only for linear indexing
 end
+# Compute the linear indexability of the indices, and combine it with the linear indexing of the parent
+function SubArray(parent::AbstractArray, indexes::Tuple, dims::Dims)
+    SubArray(linearindexing(viewindexing(indexes), linearindexing(parent)), parent, indexes, dims)
+end
+function SubArray{P, I, N}(::LinearSlow, parent::P, indexes::I, dims::NTuple{N, Int})
+    SubArray{eltype(P), N, P, I, false}(parent, indexes, dims, 0, 0)
+end
+@generated function SubArray{P, I, N}(::LinearFast, parent::P, indexes::I, dims::NTuple{N, Int})
+    # Compute the first index and stride
+    first_index = first_index_expr(:parent, :indexes, length(I.parameters))
+    stride1 = stride1expr(I.parameters, :parent, :indexes)
+    :(SubArray{eltype(P), N, P, I, true}(parent, indexes, dims, $first_index, $stride1))
+end
 
-typealias StridedArray{T,N,A<:DenseArray,I<:Tuple{Vararg{RangeIndex}}} Union{DenseArray{T,N}, SubArray{T,N,A,I}}
-typealias StridedVector{T,A<:DenseArray,I<:Tuple{Vararg{RangeIndex}}}  Union{DenseArray{T,1}, SubArray{T,1,A,I}}
-typealias StridedMatrix{T,A<:DenseArray,I<:Tuple{Vararg{RangeIndex}}}  Union{DenseArray{T,2}, SubArray{T,2,A,I}}
+# The NoSlice one-element vector type keeps dimensions without losing performance
+immutable NoSlice <: AbstractVector{Int}
```

Comment:
I'm really glad you agree that this is a better approach. Makes life _so_ much easier!